### PR TITLE
Implement various checks on the Location setters

### LINKED ISSUE #23522: Implement various checks on the Location setters
Spec: https://html.spec.whatwg.org/multipage/history.html#location

See all attributes that say "Can be set, to navigate to the given URL.". 

Code: See all the `Set_` methods at https://github.com/servo/servo/blob/7a5ce048d2e6740c91187c12abebce132f90bfaa/components/script/dom/location.rs#L78

As well as `set_url_component` which is used by those methods: https://github.com/servo/servo/blob/7a5ce048d2e6740c91187c12abebce132f90bfaa/components/script/dom/location.rs#L45

Currently failing tests(and maybe others):
- https://github.com/servo/servo/blob/master/tests/wpt/metadata/html/browsers/history/the-location-interface/location-protocol-setter.html.ini
- https://github.com/servo/servo/blob/master/tests/wpt/metadata/html/browsers/history/the-location-interface/location-protocol-setter-non-broken.html.ini
- https://github.com/servo/servo/blob/master/tests/wpt/metadata/html/browsers/history/the-location-interface/location-protocol-setter-non-broken-weird.html.ini

And the actual tests are found in https://github.com/servo/servo/tree/master/tests/wpt/web-platform-tests/html/browsers/history/the-location-interface

A start was made at https://github.com/servo/servo/pull/23368/files#diff-45df5e6ab4c1d79fae956506be147b62R244

and https://github.com/servo/servo/pull/23368/files#diff-c1196726ff21e3563c663b049214dafeL21

Note the only gotcha is that the `ServoUrl` contains an `Arc`, so you can't manipulate it directly when you do the initial parse to then perform the check, as that will actually change the url of the window. So instead you'll need to get a copy, as is for example done at https://github.com/servo/servo/pull/23368/files#diff-45df5e6ab4c1d79fae956506be147b62R244 (or you need to perform the checks without manipulating the current url)

### PR DESCRIPTION
**Pull request by utsavoza**
Since there are no existing open PRs for the issue, I am raising one based on the work done by @UxioAndrade and @braddunbar in previous two PRs for the issue (#25714 and #23670) respectively.

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #23522 
- [x] There are tests for these changes

### CODE REVIEW

File: components/script/dom/location.rs
Comment (gterzian) [inline components/script/dom/location.rs:272]: Are we doing step 5 currently? If not, I think we can do it using `as_mut_url().set_path`.

File: components/script/dom/location.rs
Comment (gterzian) [inline components/script/dom/location.rs:297]: Similar re step 5, which could be done using `set_port`.

File: components/script/dom/location.rs
Comment (gterzian) [inline components/script/dom/location.rs:361]: I think we're missing step 4 and 5 here?

File: components/script/dom/location.rs
Comment (gterzian) [inline components/script/dom/location.rs:225]: How about we add a `copy_url` argument to `set_url_component` to avoid all those double calls to `get_url`(for algorithms that need the url before calling `set_url_component`)?

**Review by gterzian — CHANGES_REQUESTED** [review]
Great work! Thanks for figuring out all those details. I think there are a few steps missing(although you could open an issue and leave that as a follow-up if there is a reason for that), and I have a small suggestion re the `set_url_component` method. 

Also, could you please squash into one commit? I think while it's a good idea to separate unrelated work into commits, in this case it is actually related and  consolidating the work into a commit would appear clearer.

File: components/script/dom/location.rs
Comment (utsavoza) [reply components/script/dom/location.rs:272]: Oh, I assumed we were executing those steps as a part of `url::quirks::set_path` through `UrlHelper::SetPathname`, but those steps seem to correspond to dom url spec.

File: components/script/dom/location.rs
Comment (utsavoza) [reply components/script/dom/location.rs:361]: Again, I presumed those steps as a part of `UrlHelper::SetSearch`.

File: components/script/dom/location.rs
Comment (utsavoza) [reply components/script/dom/location.rs:297]: There's one caveat though when setting port directly via `as_mut_url().set_port` - we'll have to manually verify and parse the input string to `Option<u16>`. However, we can use `url::quirks::set_port` to check and set the port here instead of `UrlHelper::SetPort`.

File: components/script/dom/location.rs
Comment (utsavoza) [reply components/script/dom/location.rs:225]: I think we can actually remove the method completely - the only place we might actually need this method after adding checks would be for `SetPort` method probably (as we might not directly setting port to the url there). For the rest of the setters, we can directly call the navigate method with the appropriate arguments.

**Comment by utsavoza** [comment]
> Also, could you please squash into one commit? I think while it's a good idea to separate unrelated work into commits, in this case it is actually related and consolidating the work into a commit would appear clearer.

Agreed. I'll squash into one commit once the changes are approved.

File: components/script/dom/location.rs
Comment (gterzian) [reply components/script/dom/location.rs:315]: Sounds good!

**Review by gterzian — CHANGES_REQUESTED** [review]
Looks like a big improvement in the correctness of the implementation! Just one nit left, and then after you squash I'll approve it(after which we might still have to deal with some surprise failures on the test run).

**Comment by utsavoza** [comment]
> Looks like a big improvement in the correctness of the implementation! Just one nit left, and then after you squash I'll approve it(after which we might still have to deal with some surprise failures on the test run).

Actually, I ran the wpt tests for the location interface locally. Apart from that how do I verify if we aren't breaking anything else?

**Comment by gterzian** [comment]
> Actually, I ran the wpt tests for the location interface locally. Apart from that how do I verify if we aren't breaking anything else?

We have to basically run the whole suite and see, there can also be tests that unexpectedly start passing...

**Comment by gterzian** [comment]
Thanks for the awesome work!

@bors-servo r+

**Comment by gterzian** [comment]
This one seems relevant:

```
1 unexpected results that are NOT known-intermittents:
  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function.html:
  │ FAIL [expected PASS] Set location from a function called from a parent
  │   → assert_equals: expected "http://web-platform.test:8000/html/browsers/browsing-the-web/navigating-across-documents/source/support/location-set.html" but got "http://web-platform.test:8000/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function.html"
  │ 
  │ onload</fr.onload<@http://web-platform.test:8000/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function.html:15:20
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:2042:32
```

This one does not seem relevant(I'm not sure what issue it would match by the way):

```
1 unexpected results that are NOT known-intermittents:
  ▶ Unexpected subtest result in /_mozilla/mozilla/scroll_top_null_target.html:
  │ FAIL [expected PASS] scroll_top_null_target
  │   → assert_equals: Target should be null! expected null but got Element node <a id="test"></a>
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/scroll_top_null_target.html:18:23
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  └ @http://web-platform.test:8000/_mozilla/mozilla/scroll_top_null_target.html:13:10
```

**Comment by gterzian** [comment]
Ok the first one might actually a a new intermittent, perhaps you fixed something that makes it show up, and since it seems related to naviation, it's probably best done in a follow up. So please take a look, and if it doesn't appear to point out an issue with this PR, just file an issue for it.

**Comment by gterzian** [comment]
this one seems defenitely relevant:

```
  ▶ Unexpected subtest result in /url/failure.html:
  │ FAIL [expected PASS] Location's href: http:/:@/www.example.com should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ FAIL [expected PASS] Location's href: http:@/www.example.com should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ FAIL [expected PASS] Location's href: http:/@/www.example.com should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ FAIL [expected PASS] Location's href: http:a:b@/www.example.com should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ FAIL [expected PASS] Location's href: http:/a:b@/www.example.com should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ FAIL [expected PASS] Location's href: http::@/www.example.com should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ FAIL [expected PASS] Location's href: http:@:www.example.com should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ FAIL [expected PASS] Location's href: http:/@:www.example.com should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ FAIL [expected PASS] Location's href: a should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ FAIL [expected PASS] Location's href: a/ should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ FAIL [expected PASS] Location's href: a// should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
  │ 
  │ runTests/<@http://web-platform.test:8000/url/failure.html:41:23
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  └ runTests@http://web-platform.test:8000/url/failure.html:40:10
```

File: components/script/dom/location.rs
Comment (gterzian) [inline components/script/dom/location.rs:262]: Ok I think this is the problem with `/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function.html`, the referrer should be not from the entry document but from `self.window.Document`, at least that is what I make from the test failure(so it's not a new intermittent and not actually related much to navigation)

**Comment by utsavoza** [comment]
I was able to find six relevant test failures:
1. `/url/failure.html`
2. `/html/browsers/browsing-the-web/navigating-across-documents/005.html`
3. `/html/browsers/browsing-the-web/navigating-across-documents/source/navigate-child-function.html`
4. `/html/browsers/browsing-the-web/navigating-across-documents/empty_fragment.html`
5. `/html/browsers/browsing-the-web/unloading-documents/beforeunload-synchronous.html` - `[UNEXPECTED PASS]`
6. `/fullscreen/api/document-exit-fullscreen-active-document.html` - `[UNEXPECTED PASS]`

and one NOT known-intermittent (as pointed out by @gterzian) from the logs:

```
./mach filter-intermittents wpt-jsonsummary.log --log-intermittents intermittents.log --log-filteredsummary filtered-wpt-errorsummary.log --tracker-api default --reporter-api default
1 unexpected results that are NOT known-intermittents:
  ▶ Unexpected subtest result in /_mozilla/mozilla/scroll_top_null_target.html:
  │ FAIL [expected PASS] scroll_top_null_target
  │   → assert_equals: Target should be null! expected null but got Element node <a id="test"></a>
  │
  │ @http://web-platform.test:8000/_mozilla/mozilla/scroll_top_null_target.html:18:23
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2001:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  └ @http://web-platform.test:8000/_mozilla/mozilla/scroll_top_null_target.html:13:10
```

I will try to resolve these and post any updates here.

File: components/script/dom/location.rs
Comment (utsavoza) [reply components/script/dom/location.rs:262]: Ahh, right. The referrer url shouldn't be from entry document here - I think `let referrer = Referrer::ReferrerUrl(self.get_url());` should fix this.

**Comment by jdm** [comment]
Filed #26269 for the new unrelated intermittent test result.

**Comment by gterzian** [comment]
@utsavoza awesome, sounds like it's worth updating this PR with your work and we'll give the test suite another try...

**Comment by utsavoza** [comment]
Hey @jdm, do we still require setting location.hash to empty string? I see that we did this earlier as a part of #11934 and whatwg/html#1318 to make this compliant with the spec. I found this a bit confusing since the spec does not list this step explicitly. 

In order to accommodate this change I believe the setter for hash would look something like this:
```rust
copy_url.as_mut_url().set_fragment(match value.0.as_str() {
    "" => Some("#"), // <- special case
    _ if value.0.starts_with('#') => Some(&value.0[1..]),
    _ => Some(&value.0),
});
```

**Comment by jdm** [comment]
@utsavoza The code in location.rs matches the specification (as far as I can tell), but it's not written in a way where that is clearly the case. The specification is written so that using `location.hash = ""` in JS will cause a URL parse to occur against the current location URL (but with an empty fragment value) and the new fragment value, so the resulting URL will have no fragment if an empty string was passed. We need to be sure that we preserve that behaviour.

**Comment by utsavoza** [comment]
So, I was able to fix the above mentioned failing wpt tests except for the first two.
>     1. `/url/failure.html`
The reason few tests are failing here is because in the updated implementation of `SetHref` we are now [parsing the given input value relative to the entry settings object](https://html.spec.whatwg.org/multipage/history.html#dom-location-href) as mentioned in the spec.

For example, for the given test:
```
self.test(() => {
  assert_throws_js(self[0].TypeError, () => self[0].location = test.input)
}, "Location's href: " + name)
```
The test input that is provided for example is:
```
  {
    "input": "http:/:@/www.example.com",
    "base": "about:blank",
    "failure": true
  }
```
but since we are now parsing the given input relative to entry settings object, the base url that gets picked up here is different (`http://web-platform.test:8000/url/failure.html`) from what is provided, which is now causing the test case to fail with:
```
▶ Unexpected subtest result in /url/failure.html:
  │ FAIL [expected PASS] Location's href: http:/:@/www.example.com should throw
  │   → assert_throws_js: function "() => self[0].location = test.input" did not throw
```
I am not really sure if updating the metadata for these test cases would be the correct way to go. Let me know what you think @gterzian?

<br>

>     2. `/html/browsers/browsing-the-web/navigating-across-documents/005.html`
For the second test, it seems we already have an open issue (#25065) that delineates the cause for this test failure. I have addressed this by updating the metadata for the test.

**Comment by gterzian** [comment]
> but since we are now parsing the given input relative to entry settings object, the base url that gets picked up here is different

I'm wondering if the problem is instead that somehow the call to `base_url.join(&value.0)` doesn't return an error, in the case that `base_url` is "http://web-platform.test:8000/url/failure.html" and `value.0` is "http:/:@/www.example.com"?

**Comment by utsavoza** [comment]
> If you were to `println!` the result of [this match](https://github.com/servo/servo/pull/26231/files#diff-45df5e6ab4c1d79fae956506be147b62R257), what do you see?

I did try to verify if this match would return an error or not. Apparently this code works:
```rust
let base_url = Url::parse("http://web-platform.test:8000/url/failure.html").unwrap();
match base_url.join("http:/:@/www.example.com") {
    Ok(url) => println!("url: {}", url), // -> prints "url: http://web-platform.test:8000/:@/www.example.com"
    Err(e) => eprintln!("error: {}", e)
};
```

**Comment by gterzian** [comment]
Ok so do you mean that the test expects a parsing error, while the code currently just parses the url? It might be a quirk of the underlying `url` library, I don't know. 

I would say if it cannot be debugged successfully now it might be worth updating the metadata and filing a follow-up issue, since we're overall making progress on the test suite in this PR.

**Comment by utsavoza** [comment]
> Ok so do you mean that the test expects a parsing error, while the code currently just parses the url? It might be a quirk of the underlying `url` library, I don't know.

Yes, ideally it should throw a `TypeError` but our implementation is able to parse the given input relative to base url. I am not sure though if this is a quirk of the underlying `url` lib. 

If I understand this correctly, the test case expects the href setter to fail when when trying to set the given input against the `about:blank` base url.

So I think it's the base url that gets picked up incorrectly because if the base url is `about:blank` our implementation would throw a `TypeError` as expected.

**Comment by gterzian** [comment]
Ok, I think this requires some additional research and is best done in a follow-up(also there are a bunch of test failures in that test already, and they could actually be all investigated in a follow-up)I have filed this for a follow up https://github.com/servo/servo/issues/26287

One last thing, have you checked if the behavior described at https://github.com/servo/servo/pull/26231#issuecomment-617965900 is preserved with these changes?

**Comment by utsavoza** [comment]
@gterzian hmm, I think I accidentally deleted my branch in my fork which caused this PR to be closed.

**Comment by utsavoza** [comment]
> One last thing, have you checked if the behavior described at [#26231 (comment)](https://github.com/servo/servo/pull/26231#issuecomment-617965900) is preserved with these changes?

Yes, I had updated the PR to reflect that change.

**Comment by jdm** [comment]
If there's no clear way to push the branch in a way that allows reopening the PR, go ahead and open a new one.

**Comment by utsavoza** [comment]
Thanks, re-opened the PR.

**Comment by utsavoza** [comment]
@gterzian, I think I have pushed all the changes that address the above listed failing wpt tests (and squashed the commits into one). Please have a look and let me know if the changes are good to go.

**Comment by gterzian** [comment]
Looks good, thanks! Just a quick rebase and it should be ready to go...

**Review by gterzian — APPROVED** [review]

**Comment by utsavoza** [comment]
Done!

**Comment by gterzian** [comment]
@utsavoza Thanks for completing this, it was a lot of additional work in the end!

@bors-servo r+

**Comment by gterzian** [comment]
@bors-servo retry

https://github.com/servo/servo/issues/21067
https://github.com/servo/servo/issues/26298
https://github.com/servo/servo/issues/25513
https://github.com/servo/servo/issues/26299

Plus in mac:

```
1 unexpected results that are NOT known-intermittents:
  ▶ TIMEOUT [expected OK] /_webgl/conformance2/sync/sync-webgl-specific.html
  │ 
  └ _RegisterApplication(), FAILED TO establish the default connection to the WindowServer, _CGSDefaultConnection() is NULL.
  ▶ Unexpected subtest result in /_webgl/conformance2/sync/sync-webgl-specific.html:
  └ NOTRUN [expected PASS] Overall test
```

not sure which one that is

**Comment by CYBAI** [comment]
@bors-servo retry

- #25533, #25513

**Comment by utsavoza** [comment]
> @utsavoza Thanks for completing this, it was a lot of additional work in the end!

Hey @gterzian, thanks for your guidance.

**Pull request merged**
